File: frontend/src/types/stock.ts
export type ProductStock = {
  product_stock_id: number
  product_id: number
  current_stock: number
  reserved_stock: number
  reorder_level: number
  maximum_level: number | null
  is_low_stock: boolean
  updated_at: string
}
export type StockMovement = {
  stock_movement_id: number
  product_id: number
  movement_type: string
  quantity: number
  running_balance: number
  source: string
  reference_id: number | null
  created_by: number | null
  created_at: string
}
export type StockAdjustment = {
  stock_adjustment_id: number
  product_id: number
  quantity_change: number
  reason: string
  balance_after: number
  adjusted_by: number | null
  created_at: string
}

export type NewStockAdjustment = {
  product_id: number
  quantity_change: number
  reason: string
}
export type StockBatch = {
  stock_batch_id: number
  product_id: number
  batch_number: string
  quantity: number
  expiry_date: string
  is_expired: boolean
  received_at: string
}

File: backend/app/models/stock.py
"""Inventory tables: current stock, movement ledger, manual adjustments, expiry batches."""
from datetime import datetime, date

from sqlalchemy import ForeignKey, Unicode, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class ProductStock(Base):
    __tablename__ = "ProductStock"

    product_stock_id: Mapped[int] = mapped_column(primary_key=True)
    product_id: Mapped[int]  # TODO: ForeignKey("Products.product_id") once Module 3 is merged
    current_stock: Mapped[int] = mapped_column(default=0)
    reserved_stock: Mapped[int] = mapped_column(default=0)
    reorder_level: Mapped[int] = mapped_column(default=0)
    maximum_level: Mapped[int | None]
    updated_at: Mapped[datetime] = mapped_column(server_default=func.sysutcdatetime(), onupdate=func.sysutcdatetime())


class StockMovement(Base):
    __tablename__ = "StockMovements"

    stock_movement_id: Mapped[int] = mapped_column(primary_key=True)
    product_id: Mapped[int]  # TODO: ForeignKey("Products.product_id")
    movement_type: Mapped[str] = mapped_column(Unicode(10))  # "in" or "out"
    quantity: Mapped[int]
    running_balance: Mapped[int]
    source: Mapped[str] = mapped_column(Unicode(30))  # "purchase", "sale", "adjustment"
    reference_id: Mapped[int | None]
    created_by: Mapped[int | None] = mapped_column(ForeignKey("Users.user_id"))
    created_at: Mapped[datetime] = mapped_column(server_default=func.sysutcdatetime())


class StockAdjustment(Base):
    __tablename__ = "StockAdjustments"

    stock_adjustment_id: Mapped[int] = mapped_column(primary_key=True)
    product_id: Mapped[int]  # TODO: ForeignKey("Products.product_id")
    quantity_change: Mapped[int]  # positive = increase, negative = decrease
    reason: Mapped[str] = mapped_column(Unicode(255))
    balance_after: Mapped[int]
    adjusted_by: Mapped[int | None] = mapped_column(ForeignKey("Users.user_id"))
    created_at: Mapped[datetime] = mapped_column(server_default=func.sysutcdatetime())


class StockBatch(Base):
    __tablename__ = "StockBatches"

    stock_batch_id: Mapped[int] = mapped_column(primary_key=True)
    product_id: Mapped[int]  # TODO: ForeignKey("Products.product_id")
    batch_number: Mapped[str] = mapped_column(Unicode(50))
    quantity: Mapped[int]
    expiry_date: Mapped[date]
    received_at: Mapped[datetime] = mapped_column(server_default=func.sysutcdatetime())